nussl/separation/duet.py

```diff
@@ -0,0 +1,440 @@
+        self.stft_ch0, self.stft_ch1, self.frequency_matrix = self._compute_spectrogram(self.sample_rate)
+
+        # Calculate the symmetric attenuation (alpha) and delay (delta) for each
+        # time-freq. point and return a matrix for each
+        self.symmetric_atn, self.delay = self._compute_atn_delay(self.stft_ch0, self.stft_ch1, self.frequency_matrix)
+
+        # Make histogram of attenuation-delay values and get the center values for the bins in this histogram
+        self.normalized_attenuation_delay_histogram, self.attenuation_bins, self.delay_bins = self._make_histogram()
+
+        # Find the location of peaks in the attenuation-delay plane
+        self.peak_indices = nussl.utils.find_peak_indices(self.normalized_attenuation_delay_histogram, self.num_sources,
+                                                          threshold=self.peak_threshold,
+                                                          min_dist=[self.attenuation_min_distance,
+                                                                    self.delay_min_distance])
+
+        # compute delay_peak, attenuation peak, and attenuation/delay estimates
+        self.delay_peak, atn_delay_est, self.atn_peak = self._convert_peaks()
+
+        # compute masks for separation
+        masks = self._compute_masks()
+
+        return masks
+
+    def _compute_spectrogram(self, sample_rate):
+        """ Creates the stfts matrices for channel 0 and 1, and computes the frequency matrix.
+        Parameter:
+            sample_rate (integer): sample rate
+
+        Returns:
+            stft_ch0 (np.matrix): a 2D Numpy matrix containing the stft of channel 0
+            stft_ch1 (np.matrix): a 2D Numpy matrix containing the stft of channel 1
+            wmat (np.matrix): a 2D Numpy matrix containing the frequencies of analysis of the Fourier transform
+        """
+
+        # Compute the stft of the two channel mixtures
+        self.audio_signal.stft_params = self.stft_params
+        self.audio_signal.stft()
+
+        stft_ch0 = self.audio_signal.get_stft_channel(0)
+        stft_ch1 = self.audio_signal.get_stft_channel(1)
+
+        # remove dc component to avoid dividing by zero freq. in the delay estimation
+        stft_ch0 += nussl.constants.EPSILON
+        stft_ch1 += nussl.constants.EPSILON
+
+        # Compute the freq. matrix for later use in phase calculations
+        n_time_bins = len(self.audio_signal.time_bins_vector)
+        wmat = np.array(np.tile(np.mat(self.audio_signal.freq_vector).T, (1, n_time_bins))) * (2 * np.pi / sample_rate)
+        wmat += nussl.constants.EPSILON
+        return stft_ch0, stft_ch1, wmat
+
+    @staticmethod
+    def _compute_atn_delay(stft_ch0, stft_ch1, frequency_matrix):
+        # Calculate the symmetric attenuation (alpha) and delay (delta) for each
+        # time-freq. point
+        inter_channel_ratio = (stft_ch1 + nussl.constants.EPSILON) / (stft_ch0 + nussl.constants.EPSILON)
```

Comment:
Hmmm... Looks like this happens twice (the first is up at line 192-3). I know I told you to do the first one. Can you test what happens if you get rid of the epsilons at 192-3. Just curious. We'll have a chat about it soon.